Improve nuradioreco logging module

MijnheerD opened the pull request:
Issue being addressed: Weird behaviour of logging module and unclear documentation. This should solve issue #601.

**Note** : changelog.txt has not been updated yet, is this necessary?

cg-laser reviewed (dismissed):
thanks a lot Mitja. This makes it much clearer

cg-laser commented:
There is in principle another ToDo.... We sometimes use "warning" to print out useful status information such as simulation progress. While this works fine, it is conceptually not ideal. It would be better to add a new custom log_level such as "status" that is the new default option which is between the warning and info levels.

MijnheerD commented:
Following [this ](https://stackoverflow.com/a/35804945/1691778) answer on StackOverflow, this should be pretty straightforward. I guess the most important question we have to deal with, is how exposed we want to make this to the end users (i.e. should it be easy for somebody to add their own custom logging level or not?).

cg-laser commented:
thanks for looking into the logging. This is very appreciated!
If possible, we can allow users as much flexibility as they want, but NuRadioMC should come with a good default option that works out of the box without the user having to do anything.

cg-laser commented:
I just realized that we already have a "status" log level defined in simulation.py which is used throughout simulation.py. 
```
STATUS = 31

# logging.basicConfig(format='%(asctime)s %(levelname)s:%(name)s:%(message)s')


class NuRadioMCLogger(logging.Logger):

    def status(self, msg, *args, **kwargs):
        if self.isEnabledFor(STATUS):
            self._log(STATUS, msg, args, **kwargs)


logging.setLoggerClass(NuRadioMCLogger)
logging.addLevelName(STATUS, 'STATUS')
logger = logging.getLogger("NuRadioMC")
assert isinstance(logger, NuRadioMCLogger)
```
However, other modules don't use this log level and would need to redefine it, I think. Not sure if we can define the log level at a more central stage.

cg-laser commented:
great. We also need to cover the usecase of "stand alone modules" that don't inherit from the default module such as the ray tracing class. But I guess, these scripts can still just call the "setup_looger()" function.

MijnheerD commented:
Alright, then I will just add the `addLoggingLevel()` function to the `base.module` module. Anyone can call this function in a script to add a desired logging level to the currently configured logging class (so this can also be a custom logging class).

Per default, the STATUS level will now be added when calling `setup_logger()` , which will allow any script to call `logger.status()` to print a status message. One question though @cg-laser : in your first message you said STATUS should be in between WARNING and INFO, which have values 30 and 20 respectively. But the code from `simulation.py` defines STATUS as 31, which is above WARNING. Which one do you think is better?

MijnheerD commented:
As a sidenote, I could update the code in `simulation.py` to work with the new function to add the logging level (maybe to provide some kind of example). Would it be a good idea to do this in this PR in one go?

MijnheerD commented:
Actually, the unit test with `simulation.py` now fails because the `NuRadioMCLogger` class already has the STATUS log level defined, and the `setup_logger()` tries to add it again. How should we best handle this? 
a) Remove the addition of the STATUS log level and corresponding method from `NuRadioMCLogger`
b) Not raise an AttributeError in `addLoggingLevel()` . In this case, should an attempted redefinition do nothing or overwrite the existing logging value?

fschlueter commented:
@MijnheerD, I have not been following this PR from the beginning but to answer your last question. I would remove the definition of `status` from `simulation.py`. Furthermore, `simulation.py` should use one of the two functions described in the documentation to initialise the logger.

MijnheerD commented:
@fschlueter I agree that is the cleanest solution, there is a conflict of value. The "standard" STATUS - defined in `module.py` has a value of 25, which means it sits in between WARNING and INFO. This makes sense to me, as a status update would be less critical than a warning but more in depth than a simple info message.

However, the STATUS defined in `simulation.py` has a value of 31, making it a higher level than WARNING. This would mean that when using WARNING as a log level, the STATUS messages would still be logged (which is not the case for "standard" implementation).

So there are three options here:
a) I update `simulation.py` to use the standard implementation we provide, and change the default logging level in the `simulation` class to STATUS. This would effectively yield the same behavior as the current implementation.
b) I add a new logging level to the `NuRadioMCLogger`, but it would need a different name, which implies changing all the logging calls using STATUS. Though this would be a nice example of how to implement custom logging levels.
c) Change the behavior of the `addLoggingLevel()` function to not `raise` , but maybe give the option to override the existing value.

Personally I prefer option a) , but I just to make sure this will not introduce any backward compatibility issues (I am not sure where the class is used). In this case I am also not sure whether a custom `NuRadioMCLogger` class is required.

MijnheerD commented:
I just found at least one other instance of `NuRadioMCLogger` with STATUS = 31 in `generator.py`. So maybe for the sake of consistency we can set the standard implementation of STATUS to 31 as well? Not sure if this makes sense. I also don't know if there should be a difference between the NuRadioReco and NuRadioMC loggers?

cg-laser reviewed (commented):
Hey, I'm sorry for not answering earlier. Please go ahead and implement option a). As you say, this is the best solution. Status should be between Warning and Info. 

The default logging level should be `STATUS`. Once we have this implemented, we can change all instances where we wrongly used `warning` as a log level where it should have rather been `status`. But we can do this in a separate PR. 

If you need to give the logger a base name, just name it "NuRadio" or "NuRadioLogger". Then it applies to both NuRadioReco and MC. There shouldn't be any difference between the loggers. 
But of course, the logger should tell from where the logging message came, such as the "module name" or from "simulation", or "ray tracing" or a similar meaningful name.

MijnheerD commented:
Hi all,

I went ahead and applied change (a) from my previous post to both `NuRadioMC/simulation/simulation.py` and `NuRadioMC/EvtGen/generator.py` , as these were the two files where I found mention of the `status()` logging messages. I do have to admit that it was not very clear to me what `generator.py` is supposed to be. It is not a module, right? But for the sake of consistency, I implemented its logging as if it were one. So please check if you think is correct.

I also added a note to the documentation to specify how to use the handling with NuRadioMC applications. In Python, as far as I know, loggers can inherit from each other by using their names following the module structure. So for example, a logger with the name `NuRadioReco.module`  automatically inherits from the `NuRadioReco` logger. I am pretty sure we implicitly rely on this behavior when creating the logger in our modules. To not break compatibility with NuRadioReco modules, the `module.setup_logger()` function still uses NuRadioReco as its base name. This allows NuRadioReco applications to use the function as is (ie without any arguments), but it does require NuRadioMC applications to explicitly set `name=NuRadioMC` when calling the function. While perhaps not very intuitive, I think it makes sense as the logging module is still part of the NuRadioReco folder.

If you don't like this behaviour, we could change the `setup_logger()` to return the root logger (i.e. the one without a name). This should allow all applications to use the same setup. But I am not sure how this would reflect to the already existing code.

As a sidenote, I noticed that at least some NuRadioMC modules don't use the "NuRadioMC.MODULE" convention for calling loggers. This means they will never pass their messages on to the logger created by `setup_logger()`. So either we don't consider this to be important for NuRadioMC applications, or these modules need to be adapted accordingly.

cg-laser reviewed (commented):
Thanks! I think that all looks very good. 

I'm confused about the name when initializing the logger as in `logger = module.setup_logger(name="NuRadioMC")`. In a NuRadioMC simulation, also NuRadioReco modules are called. How will the output look like in this case? 
What impact does it have to initialize a logger with a specific name? Do the names need to match in all sub-loggers to get the correct config? If this is the case, maybe just return the root logger, i.e. no name. We only ever want one logger from which all loggers should inherit their config. (I might also just not understanding how loggers work...)

It would be good to change all NuRadioReco modules that done use the proper naming scheme for loggers in the PR.

MijnheerD commented:
Hi Christian,

From my testing it indeed looks like the dot-naming scheme ensures the correct inheritance. So when the script initialises a logger with `module.setup_logger(name="NuRadioMC"), modules which have their logger defined as "NuRadioMC.MODULE" will follow the formatting as defined in the `setup_logger()` function. Other module, such as NuRadioReco modules, will still be able to use the STATUS log level (and function) - as this gets added to the logging module itself - but will be formatted using the default scheme.

When I looked at the unit test scripts, it looked like the only NuRadioReco modules which are loaded were convience ones (such as `units`). But if there is a real use case where both NRMC and NRR modules are to be loaded, calling `setup_logger(name="")` sets up and returns the root logger, at which point all modules should use the correct formatting.

Personally I would keep the default name to "NuRadioReco", and have scripts ask for the root logger explicitly when they mix both frameworks. What do you think?

cg-laser commented:
For background info: In every NuRadioMC simulation, NuRadioReco modules are executed. So having both NuRadioMC and NuRadioReco loggers is a case we need to cover. 

Correct my if I'm wrong, but I see three solutions
1) Initialize a root logger, or does this cause other problems because the root logger has special behaviour?
2) Always initialize a NuRadioMC and a NuRadioReco logger in addition to whatever the user requests. 
3) Initialize a "NuRadio" logger and expect the naming scheme to be "NuRadio.MC.modulename" or "NuRadio.Reco.MODULENAME". 

I think I prefer the first option as this is the most failsafe. E.g. in some of the example, the logger name is just "runStrawemen" and not "NuRadioMC.RunStrawmen". We could of course change that, but I think some user might always initialize a logger without knowing the naming convention... So option 1) will lead to the correct formatting in all cases.

MijnheerD commented:
Commit 34641179c14952b9170eb762d0f72f1b56fa176f basically implements the possibility to use option 1), which I have used in the later commits (like 7f50feab57bd5311b864354a7cf6a5df2c7eca18) to have the loggers of all modules to work properly.

Though one change I did, was to remove the named logger in the test scripts (like for example "runStrawemen"). I don't think I encountered any script where this logger was actually used, and with the new logging setup scripts should always use the `setup_logger()` function to get their loggers (in which it is not possible to use a name like "runStrawemen", as you need to configure the root logger). If a script really wants its logger with a custom name (which I don't really see as a use case), I guess they can always call `logging.getLogger()` to get it?

I will check out the last problems with the unit tests as soon as possible. The problem is simply that the `simulation` class uses a `logging.status()` call in its `__init__` function, which apparently sometimes gets executed before the `setup_logger()` function is called, which actually adds the `STATUS` option to the logging module.

MijnheerD commented:
@sjoerd-bouma @cg-laser All unit tests are behaving correctly with the new logging module now. Could one of you review the changes?  If everything looks good to you, I will add the information about the root logger to the documentation, and add the logging changes to the `changelog.txt` :-)

cg-laser commented:
I think this looks mostly good, but wouldn't it be better to move this utility into "NuRadioReco/utilities/logger.py"  or something similar? This would be the more natural place to look for that.

MijnheerD commented on NuRadioMC/simulation/simulation.py:
```diff
@@ -90,7 +92,7 @@ def __init__(self, inputfilename,
                  evt_time=datetime.datetime(2018, 1, 1),
                  config_file=None,
-                 log_level=logging.STATUS,
+                 log_level=LOGGING_STATUS,
```
What would happen in several unit tests, is that `simulation.py` would be imported before the `module.setup_logger()` was called. The `logging.STATUS` attribute is only defined when that function is executed. On import, the class definition would be evaluated and the test would fail because at that point `logging.STATUS` is not defined yet. Therefore, instead of forcing the `setup_logger()` to always happen before any other import statements (which is difficult to read and remember), I opted to define the global variable which can be used for this purpose.

If you have a suggestion for how to solve this differently, let me know!

cg-laser replied:
I can't think of a better solution, so I'm fine with it ;-)

MijnheerD commented:
@cg-laser Totally agree with you that it currently lives in a weird place. I would even go as far as to suggest adding a top level directory for this, as the logging functionality is shared between NRMC and NRR.

In any case, this would require all import statements to be changed, which is probably not the biggest issue but would need to be announced clearly to all users.

cg-laser commented:
Because of Python imports, it needs to live either in the NuRadioReco or the NuRadioMC folder. As NuRadioReco is more general, i.e., it also works standalone, whereas NuRadioMC always depends on NuRadioReco, we put the more general things into "NuRadioReco/utilities", and only the NuRadioMC specific things into "NuRadioMC/untilities". 

If you could do this change, that would be great. 
To alert users, we could leave the old setup_logger function and raise an error whenever it is called where it says that the function moved to utilities.

MijnheerD commented:
I moved the `setup_logger()` function to "NuRadioReco/utilities/logging.py" , but left a function with the same name in the "NuRadioReco/base/module.py" file which raises a `DeprecationWarning` and advises to update the import statement.

Note that due to the file name, it is not advised to due an import like `from NuRadioReco.utilities import logging` , as this might conflict with the standard `logging` package. The other option, calling the file "logger.py", I also did not like because most scripts initialise a logger like `logger = setup_logger()`, which would override the module name if the same style of import was used. So in the end I opted to have filename mimic the Python `logging` package, and users will have to either only import the `setup_logger()` function or import the one of the two packages under a different name.

In all the test scripts, I have opted for the first option. This is also how it is written down in the documentation.

Last point of action before we can merge, is the change log. I would suggest to put this for version 2.4.0, as to clearly mark a change in functionality for logging. Do you agree? Or shall I add it under 2.3.0-dev with the rest?

cg-laser commented:
Awesome, thanks a lot. Yes, let's better bump up the version number to 2.4

cg-laser reviewed (approved):
Thanks! I think we can merge this now.
